### DISCUSSION TITLE
Getting path parameter from RoutingContext

### DISCUSSION THREAD

**Discussion by kyngs** [quarkusio/quarkus#29368] [Q&A] [answered]
I have a method bean that automatically fetches data for a path parameter. For example:
ApiClass:
```java
@Path("v0/server/{server}/something")
public class ApiClass {
    @Inject
    ServerInfo server;
    ... methods
}
```

BeanProducer:
```java
    @Produces
    @DefaultBean
    @RequestScoped
    public ServerInfo server(RoutingContext context) {
        var name = context.pathParam("server");
        ... logic
    }
```

However, when I try to test it, the pathParam method returns null. The standard `@PathParam` annotation works correctly, but I cannot use it outside of ApiClass. Is there any way to do this?

**Comment by mschorsch** [comment]
I'm not sure if this ist the root of the problem, but ApiClass must also have the annotation @RequestScoped.

**Reply by kyngs** [reply]
> I'm not sure if this ist the root of the problem, but ApiClass must also have the annotation @RequestScoped.

Well doing so also broke the `@PathParam` annotation.

**Comment by jonjanisch** [comment]
@kyngs Have you found a solution for this? I need to accomplish the same thing.

**Reply by kyngs** [reply]
Unfortunately, I did not.

**Comment by mkouba** [comment]
I think that you're mixing two APIs - JAX-RS (`@Path`) and Vert.x Web (`RoutingContext`). The `RoutingContext` does not know anything about JAX-RS path params. I.e. `RoutingContext#pathParam()` only handles the [path params specified on a Vert.x route](https://vertx.io/docs/vertx-web/java/#_capturing_path_parameters), not a JAX-RS resource.

**Reply by kyngs** [reply]
This is very well possible as I do not really know how the internals work. I've used the `RoutingContext` before to get a cookie from the request, so I kinda assumed it would also work on a path param. What class should I use in my bean to get the path param from the JAX-RS annotation?

**Reply by mkouba** [reply]
It's ok to use the `RoutingContext` for cookies, query params, etc. But path params are a bit special.

> What class should I use in my bean to get the path param from the JAX-RS annotation?

I don't think it's possible. I mean you can inspect the JAX-RS path params within a resource with `@PathParam` or even `@RestPath` . Actually, there might be a way in RESTEasy Reactive. CC @geoand

**Reply by geoand** [reply]
Try using `jakarta.ws.rs.core.PathSegment`

**Reply by kyngs** [reply]
That unfortunately does not seem to work inside a bean. I get this error:
```
jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type jakarta.ws.rs.core.PathSegment and qualifiers
```
Even though I inject the `PathSegment` with `@PathParam`:
```java
    @Produces
    @DefaultBean
    @RequestScoped
    public redacted redacted(@PathParam("redacted") PathSegment context) {
        var name = context.getPath();
        return new redacted(-1, name, name);
    }
```

**Reply by geoand** [reply]
PathSegment is only available as a method parameter of a JAX-RS method

**Reply by kyngs** [reply]
Well yeah, but that doesn't solve my problem at all. I'm seeking a solution that would have all path params available based on the current routing context. If it can work for cookies and query params, why not for path params?

**Reply by jonjanisch** [reply]
I also encountered issues using PathSegment. In my case my needs were quite simple. I wanted a basic producer that given two  path params, creates an immutable class with both values. `@BeanParam` is close, but I would need to use some sort of converter to convert the primitive values (String/Long) into their proper type. Also, the bean I'm populating is not part of the Quarkus application and thus does not have `@BeanParam` annotation available.

For now I'm doing the workaround suggested by this SO post (https://stackoverflow.com/questions/67091769/access-path-params-from-outside-the-main-controller-with-quarkus-and-resteasy) and using a simple regex to extract the parameters from the URL. Works fine. I guess I was just expecting there would be a simpler solution.

**Reply by kyngs** [reply]
Yeah, I have had that idea too, but I was afraid that it could break so I've decided to write a bit more boilerplate and call the bean method manually every time I need it. (Which only adds a few lines per controller)
I think that an API for this would be great!

**Reply by jonjanisch** [reply]
The code looks something like this:

```
@Produces
@DefaultBean
@RequestScoped
public MyContext getMyContext(RoutingContext routingContext) {
    // Using normalizedPath, not sure if that's the best way, but it works
    Matcher matcher = pattern.matcher(routingContext.normalizedPath());  
    // some error handling
    String a = matcher.group("a");
    Long b = Long.parseLong(matcher.group("b"));
    return new MyContext(new A(a), new B(b));
}
```

**Reply by kyngs** [reply]
Can you provide the regex pattern? I probably won't use it, but it might be useful for other users if they require this functionality.

**Reply by jonjanisch** [reply]
That would be something specific for whatever URL structure their using. For us we have standardized URL, e.g.

```
/api/a/<a-id>/b/<b-id>/lots/of/subresources
```

So for now I'm using a simple regex like this:

```
\/api\/a\/(?<a>\w+)\/b\/(?<b>\d+)\/(?:.)*
```

where `?<a>` and `?<b>` are the named capture groups. This is still POC at this time. I have not merged the code into our app yet so I'll probably need to double check the regex further later.

**Reply by geoand** [reply, marked as answer]
I haven't tried it, but:

```java
    public static class PathSegmentProducer {

        @RequestScoped
        List<PathSegment> pathSegments(UriInfo uriInfo) {
            return uriInfo.getPathSegments();
        }
    }
```

should allow you to inject `List<PathSegment>` into any CDI bean.

I'll warn you though that is not really a supported use case, so if someone comes to me in the future saying it doesn't work, I can't say I'll go out of my way to fix it :)

**Reply by kyngs** [reply]
So, is there going to be a proper API implemented in the future for this? Or we will have to stick with the workaround?

**Reply by geoand** [reply]
It's unlikely there will be.
